spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/pubsub/GcpPubSubAutoConfiguration.java

```diff
@@ -76,7 +77,10 @@ public GcpPubSubAutoConfiguration(GcpPubSubProperties gcpPubSubProperties,
 		this.finalProjectIdProvider = gcpPubSubProperties.getProjectId() != null
 				? gcpPubSubProperties::getProjectId
 				: gcpProjectIdProvider;
-		this.finalCredentialsProvider = gcpPubSubProperties.getCredentials().getLocation() != null
+		Credentials credentialProperties = gcpPubSubProperties.getCredentials();
+		this.finalCredentialsProvider =
```

Comment:
Something is wrong going on here.
Would be better to have a `CredentialsProvider` as a conditional bean and inject it into all the `@Bean` method below.
With this approach you eliminate a possibility for the end-user to provide his/her own `CredentialsProvider` implementation.

This should be resolved in the separate issue though.

Reply:
@artembilan Can you please file an issue for this.

Reply:
Feel free to comment on #501.

Though to address your comment, this behaviour is desired -- credentials from properties have precedence over the `CredentialsProvider` bean.

Reply:
> credentials from properties have precedence over the `CredentialsProvider` bean.

That's not correct, especially if I would like to test auto-config against some `MockCredentialsProvider` bean in the test `@Configuration`. That's not how things work in Spring Boot. And that's why we have all those `@Conditional...`

Reply:
Fair enough! Let's continue discussion there!
Would be good theme for the next call. Today?